Sources/GRPCHTTP2Core/GRPCStreamStateMachine.swift

```diff
@@ -0,0 +1,687 @@
+      }
+      
+      guard let endpoint = metadata.endpoint else {
+        throw RPCError(code: .unimplemented, message: "No :path header has been set.")
+      }
+      
+      // TODO: Should we verify the RPCRouter can handle this endpoint here,
+      // or should we verify that in the handler?
+      
+      let encodingValues = metadata[stringValues: "grpc-encoding"]
+      var encodingValuesIterator = encodingValues.makeIterator()
+      if let rawEncoding = encodingValuesIterator.next() {
+        guard encodingValuesIterator.next() == nil else {
+          throw RPCError(
+            code: .invalidArgument,
+            message: "grpc-encoding must contain no more than one value"
+          )
+        }
+        guard let encoding = Encoding(rawValue: rawEncoding) else {
+          let status = Status(
+            code: .unimplemented,
+            message: "\(rawEncoding) compression is not supported; supported algorithms are listed in grpc-accept-encoding"
+          )
+          let trailers = Metadata(dictionaryLiteral: (
+            "grpc-accept-encoding",
+            .string(self.supportedCompressionAlgorithms
+              .map({ $0.rawValue })
+              .joined(separator: ",")
+            )
+          ))
+          return .reject(status: status, trailers: trailers)
+        }
+
+        guard self.supportedCompressionAlgorithms.contains(where: { $0 == encoding }) else {
+          if self.supportedCompressionAlgorithms.isEmpty {
+            throw RPCError(
+              code: .unimplemented,
+              message: "Compression is not supported"
+            )
+          } else {
+            let status = Status(
+              code: .unimplemented,
+              message: "\(encoding) compression is not supported; supported algorithms are listed in grpc-accept-encoding"
+            )
+            let trailers = Metadata(dictionaryLiteral: (
+              "grpc-accept-encoding",
+              .string(self.supportedCompressionAlgorithms
+                .map({ $0.rawValue })
+                .joined(separator: ",")
+              )
+            ))
+            return .reject(status: status, trailers: trailers)
+          }
+        }
+        
+        // All good
+      }
+
+      switch self.state {
+      case .clientIdleServerIdle(let state):
+        self.state = .clientOpenServerIdle(.init(
+          previousState: state,
+          compressionAlgorithm: metadata.encoding
+        ))
+        return .doNothing
+      case .clientOpenServerIdle, .clientOpenServerOpen, .clientOpenServerClosed:
+        throw assertionFailureAndCreateRPCErrorOnFailedPrecondition("Client shouldn't have sent metadata twice.")
+      case .clientClosedServerIdle, .clientClosedServerOpen, .clientClosedServerClosed:
+        throw assertionFailureAndCreateRPCErrorOnFailedPrecondition("Client can't have sent metadata if closed.")
+      }
+    }
+    
+    mutating func receive(message: ByteBuffer, endStream: Bool) throws {
```

Comment:
nit: at this point we don't know it's a message (it may be part of a message, multiple messages, etc.). It might be clearer to call this `bytes`/`buffer`/`data` something along those lines to make it clear it hasn't yet been de-framed.